src/Gui/ExpressionCompleter.cpp

```diff
@@ -17,317 +18,1973 @@
+            QString res;
+            if(row & 1)
+                res = QString::fromUtf8(quote(doc->Label.getStrValue()).c_str());
+            else
+                res = QString::fromLatin1(doc->getName());
+            if(sep)
+                res += QLatin1String("#");
+            return res;
+        }
+
+        static QString objName(App::DocumentObject *obj, int row, bool sep=true) {
+            QString res;
+            if(!obj || !obj->getNameInDocument())
+                return res;
+            if(sep)
+                res = QLatin1String(".");
+            if(row & 1)
+                res += QString::fromUtf8(quote(obj->Label.getStrValue()).c_str());
+            else
+                res += QString::fromLatin1(obj->getNameInDocument());
+            return res;
+        }
+
+        static QVariant docData(App::Document *doc, int row, int role) {
+            static QIcon icon(Gui::BitmapFactory().pixmap("Document"));
+            switch(role) {
+            case Qt::UserRole:
+            case Qt::EditRole:
+                return docName(doc, row, true);
+            case Qt::DisplayRole:
+                return docName(doc, row, false);
+            case Qt::DecorationRole:
+                return icon;
+            default:
+                return QVariant();
+            }
+        }
+
+        QVariant objData(App::DocumentObject *obj,
+                int row, int role, bool local=false, bool sep=true) const
+        {
+            switch(role) {
+            case Qt::UserRole:
+            case Qt::EditRole:
+                return objName(obj, row, sep);
+            case Qt::DisplayRole: {
+                QString name = objName(obj, row, local);
+                if(getModel()->inList.count(obj))
+                    name += QObject::tr(" (Cyclic reference!)");
+                return name;
+            }
+            case Qt::DecorationRole: {
+                auto vp = Gui::Application::Instance->getViewProvider(obj);
+                if(vp)
+                    return vp->getIcon();
+                return QIcon();
+            }
+            case Qt::ForegroundRole:
+                if(getModel()->inList.count(obj))
+                    return QColor(255.0, 0, 0);
+                break;
+            default:
+                break;
+            }
+            return QVariant();
+        }
+
+        QVariant sobjData(App::DocumentObject *obj, App::DocumentObject *sobj,
+                int row, int role, bool local=false) const
+        {
+            if(obj && obj->getNameInDocument()
+                    && sobj && sobj->getNameInDocument()
+                    && !(row & 1))
+            {
+                if(role == Qt::EditRole)
+                    return QString::fromLatin1(".%1.").arg(
+                        QString::fromLatin1(sobj->getNameInDocument()));
+                else if(role == Qt::UserRole && obj->getPropertyByName(sobj->getNameInDocument())) {
```

Comment:
I think it might be a good idea to give Qt::UserRole (and Qt::UserRole+1, which I think is also used later on) more descriptive names within this file, either via a locally-defined enumeration, or just some const ints.

Reply:
Fair point. I'll change that later